src/librustc_span/symbol.rs

```diff
@@ -1438,6 +1540,51 @@ impl Symbol {
     }
 }
 
+// These `Symbol` methods are for accessing symbols via `SESSION_GLOBALS`.
+impl Symbol {
+    /// Map a string to its symbol representation, using the interner's table
+    /// from `SESSION_GLOBALS` if necessary (a relatively slow operation).
+    pub fn intern(string: &str) -> Self {
+        if let Some(sym) = Symbol::try_new_inlined(string) {
+            sym
+        } else {
+            with_interner(|interner| interner.intern(string))
+        }
+    }
+
+    /// Access the symbol's chars, using the interner's table from
+    /// `SESSION_GLOBALS` if necessary (a relatively slow operation).
+    pub fn with<F: FnOnce(&str) -> R, R>(self, f: F) -> R {
+        f(self.as_str().deref())
+    }
+
+    /// Convert to a `SymbolStr`, using the interner's table from
+    /// `SESSION_GLOBALS` if necessary (a relatively slow operation).
+    pub fn as_str(self) -> SymbolStr {
```

Comment:
Future idea: make this function act on `&self` and return a `&str`. This way we can get rid of the `SymbolStr` type entirely and it would address the "Leaking a symbol's `&'static str` beyond the lifetime of the interner" issue. This will likely require major changes in rustc though, as `'static` is very convenient.

Reply:
Getting rid of `SymbolStr` would indeed be nice. I have written this idea down on my todo list.

Reply:
I just tried this out, it's intriguing.

First, were you thinking that `as_str()` would become something like this?
```
    pub fn as_str(&self) -> &str {
        with_interner(|interner| unsafe {
            std::mem::transmute::<&str, &str>(interner.get(*self))
        })
    }
```
Instead of falsely claiming that the lifetime of the result is `static`, it falsely claims that the lifetime matches that of `&self`. So it's still a lie, but it's a much smaller lie. (Unless I've misunderstood something?)

I tried making this change without this PR's commits applied. It worked amazingly well. There are only about a dozen places where I need to use the old version, mostly where we are sorting vectors containing symbols, e.g.:
```
        names.sort_by_key(|n| n.as_str());
```
(I get borrowck errors with the new `as_str()` formulation.) The `ToStableHashKey` impl for `Symbol` is also tricky.

I can imagine with some effort it might be possible to eliminate `SymbolStr` entirely. Also, working with `&str` is nicer than `SymbolStr`, because it avoids the need for some `&` and `&*` occurrences.

Having said all that, won't this PR block this new approach? `SymbolStr` must exist with inline symbols -- because it contains the byte array in which we de-inline inlined symbols for `as_str()`.

Reply:
Unless I could remove the byte array and just make the `&str` point to the symbol's u32. It would work on little-endian machines, I'd have to think about what would be necessary for it to work on big-endian.

Reply:
Aha, I can fix the sort problems by switching from `sort_by_key` to `sort_by`.

Reply:
Yea, I think we'll be able to find a (non-allocating) solution for all use-sites, even if they may require some refactorings.

> Unless I could remove the byte array and just make the `&str` point to the symbol's u32. It would work on little-endian machines, I'd have to think about what would be necessary for it to work on big-endian.

Yea that's what I was thinking, although I didn't realize endianess could be a problem. Could we just encode the bytes independently of the endianness, so the bytes are always correct? Or will that mess with this PRs indexing scheme too much?

Reply:
I don't think we should block this PR on it. This PR should be merged regardless of whether we will be able to implement the above scheme.

Reply:
I ended up removing `SymbolStr` in #91957.